Revert "Update plan-specific flows to use generic plans step"

andrewserong opened the pull request:
Reverts Automattic/wp-calypso#35047

Now that https://github.com/Automattic/wp-calypso/pull/35109 has been merged (remove index check in `removeUnneededSteps` we can attempt to revert the temporary fix for the broken transition to the ecommerce flow: https://github.com/Automattic/wp-calypso/issues/32506

## Testing

We wish to test that all of the plan-specific flows function as they used to. Pay particular attention to using the 'back' button and that it takes you to the correct step.

* From an incognito window, go to http://calypso.localhost:3000/start/personal
* Log in on the user step
* From the Site type step, select Online store, and it should take you to the domains step on the ecommerce-onboarding flow, and you should see: http://calypso.localhost:3000/start/ecommerce-onboarding/domains in the address bar
* Click back, and you should be on the site-type step
* Select any other site type (e.g. blog) and you should be on the site-topic-with-preview step (e.g. http://calypso.localhost:3000/start/personal/site-topic-with-preview )
* Continue on with the signup flow

Repeat the above steps for:

* http://calypso.localhost:3000/start/premium
* http://calypso.localhost:3000/start/business
* http://calypso.localhost:3000/start/free
* http://calypso.localhost:3000/start/

For each flow, ensure that a site can be created, or that you arrive on the correct checkout with the correct plan pre-selected.

andrewserong committed:
Revert "Update plan-specific flows to use generic plans step (#35047)"

This reverts commit edf16d95c1cc3fcf91043dbf58939db8247fc853.

andrewserong committed:
Update lastModified dates for plan specific flows

michaeldcain commented:
This might have existed before #35109, but if I go to /start/premium or /start/business, choose the "online store" segment, and make it all the way to the plans page. Then go back to the site type step, choose a different segment, and finish Signup, my cart is empty when I'm directed to checkout.

ramonjd commented:
Just FYI with the index check in #35109 reinstated, I can't return to the site type step from ecommerce:

![Aug-05-2019 11-16-45](https://user-images.githubusercontent.com/6458278/62432375-ab52dd00-b772-11e9-80cc-c63d452567d7.gif)

ramonjd commented:
@andrewserong @michaeldcain 

> This might have existed before #35109, but if I go to /start/premium or /start/business, choose the "online store" segment, and make it all the way to the plans page. Then go back to the site type step, choose a different segment, and finish Signup, my cart is empty when I'm directed to checkout. 

Though I haven't yet tested this inference, I believe that this probably did exist before hand.

What I'm seeing is that, when I land on `/start/business` flow, `state.signup.progress` is populated with a cart item

<img width="553" alt="Screen Shot 2019-08-05 at 11 52 17 am" src="https://user-images.githubusercontent.com/6458278/62433470-dd1a7280-b777-11e9-98c1-992e461ecb01.png">

Changing flows to `ecommerce-onboarding` however, wipes that _'unneeded'_ step, so, when you return to the site type step your progress will look like:

<img width="674" alt="Screen Shot 2019-08-05 at 11 56 14 am" src="https://user-images.githubusercontent.com/6458278/62433511-0dfaa780-b778-11e9-8e3a-c0673efbb810.png">

So, my guess so far (still 👀 ) is that we'd need to add the step again...

andrewserong commented:
Good find @ramonjd — I was only looking in the `dependencyStore` so I couldn't find anything amiss, but that gets to the heart of it. It's almost like when someone clicks 'back' and switches flow in the process, we might want it to refresh / effectively reload the flow as if it was a fresh pageload? I'll keep exploring, too...

ramonjd commented:
>  I was only looking in the dependencyStore so I couldn't find anything amiss,

Honing in, we're performing a couple of step in main.jsx – `removeFulfilledSteps()` and `processFulfilledSteps()`. Ultimately the step method `fulfilledStepCallback()` is called (which is mapped in steps-pure to `isPlanFulfilled` in step-actions.js. 

It's here that were reporting that the plan is fulfilled even though it's wiped from the progress store.

andrewserong committed:
Add `resetExcludedSteps` method when changing flow name

This allows the flow to be 'refreshed' from its original set of steps when switching flows, instead of pre-excluding steps.

andrewserong committed:
Add resetExcludedSteps method to flows mock for flow-controller tests

andrewserong committed:
Revert `resetExcludedSteps` to remove the change from this branch

To prevent this branch from becoming more complicated, I'll continue exploring `resetExcludedSteps` on a separate branch, keeping this for just reverting the plan-specific flows back to their origin state.

spen commented:
> This might have existed before #35109, but if I go to /start/premium or /start/business, choose the "online store" segment, and make it all the way to the plans page. Then go back to the site type step, choose a different segment, and finish Signup, my cart is empty when I'm directed to checkout.

I may be wrong, but it looks like this could be solved already.
IIRC those flows add a progress step that contains the cart item, when we switch to another flow we get rid of that data so that would explain the issue. On an older branch I could see this was the case but now I've pulled the latest changes it seems to work fine :)

ramonjd commented:
> On an older branch I could see this was the case but now I've pulled the latest changes it seems to work fine :)

@spen Did you do anything special to test this? I'm re-testing this branch, and can still see the issue.

ramonjd reviewed (approved):
I just rebased to kick off the tests.

In my view, this revert LGTM

The 'changing flows' bug is something we need to fix, but getting the plan flows back to normal I believe takes higher priority, especially given that the bug seems to have been loitering around for some time.

andrewserong commented:
Thanks @ramonjd — I agree. I'm happy to merge this once the e2e tests are passing.

spen commented:
> Spen Did you do anything special to test this? I'm re-testing this branch, and can still see the issue.

Nothing special, I was just logging the progress steps and then checking that the correct products were in the checkout 👍
